File: plugin/server/routes/plugins.py
"""
插件管理路由
"""
from typing import Optional

from fastapi import APIRouter, Query

from plugin.logging_config import get_logger
from plugin.server.application.plugins import PluginLifecycleService, PluginQueryService
from plugin.server.domain.errors import ServerDomainError
from plugin.server.infrastructure.auth import require_admin
from plugin.server.infrastructure.error_mapping import raise_http_from_domain

router = APIRouter()
logger = get_logger("server.routes.plugins")
query_service = PluginQueryService()
lifecycle_service = PluginLifecycleService()


@router.get("/plugin/status")
async def plugin_status(plugin_id: Optional[str] = Query(default=None)) -> dict[str, object]:
    try:
        return await query_service.get_plugin_status(plugin_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.get("/plugins")
async def list_plugins() -> dict[str, object]:
    try:
        return await query_service.list_plugins()
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin/{plugin_id}/start")
async def start_plugin_endpoint(plugin_id: str, _: str = require_admin) -> dict[str, object]:
    try:
        return await lifecycle_service.start_plugin(plugin_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin/{plugin_id}/stop")
async def stop_plugin_endpoint(plugin_id: str, _: str = require_admin) -> dict[str, object]:
    try:
        return await lifecycle_service.stop_plugin(plugin_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin/{plugin_id}/reload")
async def reload_plugin_endpoint(plugin_id: str, _: str = require_admin) -> dict[str, object]:
    try:
        return await lifecycle_service.reload_plugin(plugin_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugins/reload")
async def reload_all_plugins_endpoint(_: str = require_admin) -> dict[str, object]:
    """
    重载所有插件
    
    停止所有运行中的插件，然后重新加载。
    用于前端全局重载按钮。
    """
    try:
        return await lifecycle_service.reload_all_plugins()
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin/{ext_id}/extension/disable")
async def disable_extension_endpoint(ext_id: str, _: str = require_admin) -> dict[str, object]:
    """禁用 Extension：通知宿主进程卸载 Router"""
    try:
        return await lifecycle_service.disable_extension(ext_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)


@router.post("/plugin/{ext_id}/extension/enable")
async def enable_extension_endpoint(ext_id: str, _: str = require_admin) -> dict[str, object]:
    """启用 Extension：通知宿主进程重新注入 Router"""
    try:
        return await lifecycle_service.enable_extension(ext_id)
    except ServerDomainError as error:
        raise_http_from_domain(error, logger=logger)

File: plugin/frontend/vue-project/src/api/plugins.ts
/**
 * 插件相关 API
 */
import { get, post } from './index'
import type {
  PluginMeta,
  PluginStatusData,
  PluginHealth,
  PluginMessage,
} from '@/types/api'

/**
 * 获取插件列表
 */
export function getPlugins(): Promise<{ plugins: PluginMeta[]; message: string }> {
  return get('/plugins')
}

/**
 * 获取插件状态
 */
export function getPluginStatus(pluginId?: string): Promise<PluginStatusData | { plugins: Record<string, PluginStatusData> }> {
  const url = pluginId ? `/plugin/status?plugin_id=${encodeURIComponent(pluginId)}` : '/plugin/status'
  return get(url)
}

/**
 * 获取插件健康状态
 */
export function getPluginHealth(pluginId: string): Promise<PluginHealth> {
  const safeId = encodeURIComponent(pluginId)
  return get(`/plugin/${safeId}/health`)
}

/**
 * 启动插件
 */
export function startPlugin(pluginId: string): Promise<{ success: boolean; plugin_id: string; message: string }> {
  const safeId = encodeURIComponent(pluginId)
  return post(`/plugin/${safeId}/start`)
}

/**
 * 停止插件
 */
export function stopPlugin(pluginId: string): Promise<{ success: boolean; plugin_id: string; message: string }> {
  const safeId = encodeURIComponent(pluginId)
  return post(`/plugin/${safeId}/stop`)
}

/**
 * 重载插件
 */
export function reloadPlugin(pluginId: string): Promise<{ success: boolean; plugin_id: string; message: string }> {
  const safeId = encodeURIComponent(pluginId)
  return post(`/plugin/${safeId}/reload`)
}

/**
 * 重载所有插件（批量 API，后端并行处理）
 */
export function reloadAllPlugins(): Promise<{
  success: boolean
  reloaded: string[]
  failed: { plugin_id: string; error: string }[]
  skipped: string[]
  message: string
}> {
  return post('/plugins/reload')
}

/**
 * 获取插件消息
 */
export function getPluginMessages(params?: {
  plugin_id?: string
  max_count?: number
  priority_min?: number
}): Promise<{ messages: PluginMessage[]; count: number; time: string }> {
  return get('/plugin/messages', { params })
}

/**
 * 禁用 Extension（热切换）
 */
export function disableExtension(extId: string): Promise<{ success: boolean; ext_id: string; host_plugin_id: string; data?: any; message?: string }> {
  const safeId = encodeURIComponent(extId)
  return post(`/plugin/${safeId}/extension/disable`)
}

/**
 * 启用 Extension（热切换）
 */
export function enableExtension(extId: string): Promise<{ success: boolean; ext_id: string; host_plugin_id: string; data?: any; message?: string }> {
  const safeId = encodeURIComponent(extId)
  return post(`/plugin/${safeId}/extension/enable`)
}

/**
 * 获取服务器信息（包括SDK版本）
 */
export function getServerInfo(): Promise<{
  sdk_version: string
  plugins_count: number
  time: string
}> {
  return get('/server/info')
}

